### PR TITLE
[opt-remark] Add a new SIL testing option that works around not having decls in SIL for debug_value.

### PR DESCRIPTION
**Pull request by gottesmm**
Instead, in such a case, we use the name on the debug_value instead and just use
as the loc the debug_value itself. This will let me write SIL test cases for
opt-remark-gen.

I am going to add SIL test cases for future changes (as well as swift ones)
using this technique. I am going to in forthcoming commits fill in some tests
for the current opt-remark-generation here. I just want to get in this part.

@adrian-prantl this is what we spoke about. I finally got around to doing it.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
Actually found a small mistake. I am going to add a little more on top of this that fixes the problem.

File: lib/SILOptimizer/Transforms/OptRemarkGenerator.cpp [hunk 1]
```diff
@@ -111,8 +117,14 @@ void ValueToDeclInferrer::printNote(llvm::raw_string_ostream &stream,
 
     llvm_unreachable("Covered switch is not covered?!");
   }
+}
 
-  accessPath.clear();
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/OptRemarkGenerator.cpp:115]: NOTE: I clear in a SWIFT_DEFER at the top of ValueToDeclInferrer::infer

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SILOptimizer/Transforms/OptRemarkGenerator.cpp [hunk 2]
```diff
@@ -41,6 +41,15 @@ static llvm::cl::opt<bool> ForceVisitImplicitAutogeneratedFunctions(
         "Emit opt remarks even on implicit and autogenerated functions"),
     llvm::cl::init(false));
 
+static llvm::cl::opt<bool> DecllessDebugValueUseSILDebugInfo(
+    "optremarkgen-declless-debugvalue-use-sildebugvar-info", llvm::cl::Hidden,
+    llvm::cl::desc(
+        "If a debug_value does not have a decl, infer a value with a name from "
+        "that info that has a loc set to the loc of the debug_value "
+        "instruction itself. This is for testing purposes so it is easier to "
+        "write SIL test cases for this pass"),
```
Comment (adrian-prantl) [inline lib/SILOptimizer/Transforms/OptRemarkGenerator.cpp:50]: This doesn't look quite right to me. You are not supposed to extaract the variable name from the Decl. All instruction that carry variable debug info (`debug_value`, `debug_value_addr`, `alloc_stack`, `alloc_box`, ...?) have a `getVarInfo()` method that should be used instead.

Comment (adrian-prantl) [reply lib/SILOptimizer/Transforms/OptRemarkGenerator.cpp:50]: see https://github.com/apple/swift/blob/bf7c03b1f6cce062e27684abe5018a723a594662/include/swift/SIL/SILInstruction.h#L1467 for example.

Comment (gottesmm) [reply lib/SILOptimizer/Transforms/OptRemarkGenerator.cpp:50]: Ok. Let me fix that.

Comment (gottesmm) [reply lib/SILOptimizer/Transforms/OptRemarkGenerator.cpp:50]: I think the option is still needed since the loc that I need is from the decl.

**Pull request merged**